### PR TITLE
[WIP] New version, multiobjective optimization

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

New version, after the massive failure of my attempt in the "moo" branch

## How Has This Been Tested (if it applies)

i have not yet reintroduced the tests, I'm doing so now :-)

## Checklist

comments below for *after* I will have finished reintroducing my changes...

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/functions/multiobjective.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+            return -my_hypervolume
+
+        # Now we compute for each axis 
+        distance_to_pareto = float("Inf")
+        for y in self.pointset.values():
+            if all(y <= v):
+                distance_to_pareto = min(distance_to_pareto, min(v-y))
+        return -my_hypervolume + distance_to_pareto
+
+    def __call__(self, candidate: Candidate):
```
Comment (jrapin) [inline nevergrad/functions/multiobjective.py]: to make it more transparent, this (and other functions) should not take Candidate as input, just *args and **kwargs

File: nevergrad/functions/multiobjective.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+def hypervolume(pointset, ref):
+    """Compute the absolute hypervolume of a *pointset* according to the
+    reference point *ref*.
+    """
+    hv = _HyperVolume(ref)
+    return hv.compute(pointset)
+
+
+class MultiobjectiveFunction(InstrumentedFunction):
```
Comment (jrapin) [inline nevergrad/functions/multiobjective.py]: This should not need to derive from InstrumentedFunction

File: nevergrad/functions/multiobjective.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+from typing import (Tuple, Any, Callable, List, Optional, Dict, ValuesView, Iterator,
+                    TypeVar, Generic, Deque, Iterable)
+import warnings
+import numpy
+#from ..optimization import base
+from ..optimization.base import Candidate
+from ..common.typetools import ArrayLike
+from ..instrumentation import InstrumentedFunction
+
+def hypervolume(pointset, ref):
```
Comment (jrapin) [inline nevergrad/functions/multiobjective.py]: anything that you have added should be mode to a different file:
I would go with a directiory:
multiobjective
   __init__.py
   hypercube.py (or the initial name of the file)
   core.py (our file)

then we'll add files with examples

File: nevergrad/functions/multiobjective.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+    """
+
+
+    def __init__(self, multiobjective_function: Callable[..., Tuple[float, ...]], upper_bounds: Tuple[float, ...]) -> None:
+        self.multi_objective_function = multiobjective_function
+        self.upper_bounds = numpy.array(upper_bounds)
+        self.pointset: Dict[Candidate, ArrayLike] = {}
+        self.best_hypervolume = -float("Inf")
+
+    def compute_aggregate_loss(self, x: Candidate, v: Tuple[float]) -> None:
```
Comment (jrapin) [inline nevergrad/functions/multiobjective.py]: what is v?? losses? the name is not explicit enough
the signature should be `losses: Tuple[float,...], *args: Any, **kwargs: Any`

Comment (teytaud) [reply nevergrad/functions/multiobjective.py]: v is the objective function values, i.e. the losses

**Pull request closed without merging**